src/abstract_charm.py

```diff
@@ -0,0 +1,343 @@
+        cos_relation_exists = (
+            self._cos_relation.relation_exists
+            and not self._cos_relation.is_relation_breaking(event)
+        )
+        if cos_relation_exists:
+            return self._cos_relation.exporter_user_config
+
+    def get_workload(self, *, event):
+        """MySQL Router workload"""
+        if connection_info := self._database_requires.get_connection_info(event=event):
+            return self._authenticated_workload_type(
+                container_=self._container,
+                logrotate_=self._logrotate,
+                connection_info=connection_info,
+                cos=self._cos_relation,
+                charm_=self,
+            )
+        return self._workload_type(
+            container_=self._container, logrotate_=self._logrotate, cos=self._cos_relation
+        )
+
+    @staticmethod
+    # TODO python3.10 min version: Use `list` instead of `typing.List`
+    def _prioritize_statuses(statuses: typing.List[ops.StatusBase]) -> ops.StatusBase:
+        """Report the highest priority status.
+
+        (Statuses of the same type are reported in the order they were added to `statuses`)
+        """
+        status_priority = (
+            ops.BlockedStatus,
+            ops.MaintenanceStatus,
+            ops.WaitingStatus,
+            # Catch any unknown status type
+            ops.StatusBase,
+        )
+        for status_type in status_priority:
+            for status in statuses:
+                if isinstance(status, status_type):
+                    return status
+        return ops.ActiveStatus()
+
+    def _determine_app_status(self, *, event) -> ops.StatusBase:
+        """Report app status."""
+        if self._upgrade and (upgrade_status := self._upgrade.app_status):
+            # Upgrade status should take priority over relation status—even if the status level is
+            # normally lower priority.
+            # (Relations should not be modified during upgrade.)
+            return upgrade_status
+        statuses = []
+        for endpoint in (self._database_requires, self._database_provides):
+            if status := endpoint.get_status(event):
+                statuses.append(status)
+        return self._prioritize_statuses(statuses)
+
+    def _determine_unit_status(self, *, event) -> ops.StatusBase:
+        """Report unit status."""
+        statuses = []
+        workload_status = self.get_workload(event=event).status
+        if self._upgrade:
+            statuses.append(self._upgrade.get_unit_juju_status(workload_status=workload_status))
+        statuses.append(workload_status)
+        return self._prioritize_statuses(statuses)
+
+    def set_status(self, *, event, app=True, unit=True) -> None:
+        """Set charm status."""
+        if app and self._unit_lifecycle.authorized_leader:
+            self.app.status = self._determine_app_status(event=event)
+            logger.debug(f"Set app status to {self.app.status}")
```

Comment:
Regarding the security risk, I agree that the chances of being vulnerable is quite low, and I did not see any instances that would be problematic for that in this codebase.

For the performance impact, it can really depend on exactly what is being templated. Just using variables that are already strings or that have a simple `__str__` implementation would not have a big impact, but it can enable calling more expensive code. I think the performance argument is not great, but for efficiency it can help in the long run (particularly in the case of the debug level).

Another argument for delegating the formatting to the logger not discussed in the style guide is that it enables it to send structured logs, or add labels based on the content. I do not think this is a big concern in charms, but it might enable interesting things in the future with cos. In your example, the log could be labeled per database, helping searches.

In any case, I was just pointing out what the style guide recommends and your team is better placed than me to decide to follow it or not.